kani-compiler/src/codegen_cprover_gotoc/codegen/typ.rs

```diff
@@ -804,13 +842,181 @@ impl<'tcx> GotocCtx<'tcx> {
+    ///     Suspend0 { b: &bool, a: bool },   // state after suspending (`yield`ing) for the first time
+    ///     Suspend1,                         // state after suspending (`yield`ing) for the second time
+    /// }
+    /// ```
+    ///
+    /// However, its layout may differ from normal Rust enums in the following ways:
+    /// * Contrary to enums, the discriminant may not be at offset 0.
+    /// * Contrary to enums, there may be other fields than the discriminant "at the top level" (outside the variants).
+    ///
+    /// This means that we CANNOT use the enum translation, which would be roughly as follows:
+    ///
+    /// ```ignore
+    /// struct GeneratorEnum {
+    ///     int case; // discriminant
+    ///     union GeneratorEnum-union cases; // variant
+    /// }
+    ///
+    /// union GeneratorEnum-union {
+    ///     struct Unresumed variant0;
+    ///     struct Returned variant1;
+    ///     // ...
+    /// }
+    /// ```
+    ///
+    /// Instead, we use the following translation:
+    ///
+    /// ```ignore
+    /// union GeneratorEnum {
+    ///     struct DirectFields direct_fields;
+    ///     struct Unresumed generator_variant_Unresumed;
+    ///     struct Returned generator_variant_Returned;
+    ///     // ...
+    /// }
+    ///
+    /// struct DirectFields {
+    ///     // padding (for bool *b in Suspend0 below)
+    ///     char case;
+    ///     // padding (for bool a in Suspend0 below)
+    /// }
+    ///
+    /// struct Unresumed {
+    ///     // padding (this variant has no fields)
+    /// }
+    ///
+    /// // ...
+    ///
+    /// struct Suspend0 {
+    ///     bool *generator_field_0; // variable b in the generator code above
+    ///     // padding (for char case in DirectFields)
+    ///     bool generator_field_1; // variable a in the generator code above
+    /// }
+    /// ```
+    ///
+    /// Of course, if the generator has any other top-level/direct fields, they'd be included in the `DirectFields` struct as well.
+    fn codegen_ty_generator(&mut self, ty: Ty<'tcx>) -> Type {
+        let generator_name = self.ty_mangled_name(ty);
+        let pretty_name = self.ty_pretty_name(ty);
+        let generator_enum = self.ensure_union(self.ty_mangled_name(ty), pretty_name, |ctx, _| {
+            let type_and_layout = ctx.layout_of(ty);
+            let (discriminant_field, variants) = match &type_and_layout.variants {
+                Variants::Multiple { tag_encoding, tag_field, variants, .. } => {
+                    assert!(matches!(tag_encoding, TagEncoding::Direct));
+                    (tag_field, variants)
+                }
+                Variants::Single { .. } => {
+                    unreachable!("generators cannot have a single variant")
+                }
+            };
+            let mut fields = vec![];
+            // generate a struct for the direct fields of the layout (fields that don't occur in the variants)
+            fields.push(DatatypeComponent::Field {
```

Comment:
Can combine this by putting the first field into the vec

Reply:
Done